### PR TITLE
Feat/crud rooms rentals services

### PR DESCRIPTION
**Pull request by maironald**
# Description

- CRUD features in rooms
- CRUD features in rentals
- CRUD features in services

# Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings

# Screenshots

- Table rooms:

Screenshot from 2024-01-12 13-43-20

- Table renters:

Screenshot from 2024-01-12 13-54-20

- Table services:

Screenshot from 2024-01-12 13-55-43

- UI Dashboard

Screenshot from 2024-01-12 13-56-31

### CODE REVIEW

File: app/controllers/renters_controller.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  def show; end
+
+  def new
+    @renter = Renter.new
+    @renter.rooms.build
+    @default_rooms = current_user.rooms.select { |room| room.renter_id.nil? }
+  end
+
+  def edit
+    @default_rooms = current_user.rooms.select { |room| room.renter_id.nil? }
```
Comment (jacky-le-goldenowl) [inline app/controllers/renters_controller.rb:21]: Same above

File: app/controllers/renters_controller.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  end
+
+  def create
+    @renter = Renter.new(mapped_renter_params)
+    @room_info = Room.find_by(id: renter_params[:rooms_attributes]['0'][:id])
```
Comment (jacky-le-goldenowl) [inline app/controllers/renters_controller.rb:26]: renter có nhiều phòng à?
Chỗ này cần làm lại

File: app/controllers/renters_controller.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    if @renter.save
+      Room.where(id: @room_info.id).update(renter_id: @renter.id)
```
Comment (jacky-le-goldenowl) [inline app/controllers/renters_controller.rb:29]: Làm lại nguyên controller này

File: app/controllers/timesheets_controller.rb [hunk 4]
```diff
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 class TimesheetsController < ApplicationController
+  # The before_action is used to avoid duplicating the same code in multiple places and ensure that the necessary setup is done consistently before executing those actions (4 actions: show, edit, update, destroy)
```
Comment (jacky-le-goldenowl) [inline app/controllers/timesheets_controller.rb:4]: remove this

File: app/javascript/controllers/hello_controller.js [hunk 5]
```diff
@@ -2,7 +2,7 @@ import { Controller } from '@hotwired/stimulus';
 
 export default class extends Controller {
   connect() {
-    debugger;
+    // debugger;
```
Comment (jacky-le-goldenowl) [inline app/javascript/controllers/hello_controller.js:5]: remove this line

File: app/javascript/controllers/shared/flash_controller.js [hunk 6]
```diff
@@ -5,7 +5,13 @@ export default class extends Controller {
 
   connect() {
     setTimeout(() => {
-      this.alertTarget.close();
+      this.closeAlert();
     }, 3000);
   }
+  closeAlert() {
+    // Set display to "none" or remove the element from the DOM
```
Comment (jacky-le-goldenowl) [inline app/javascript/controllers/shared/flash_controller.js:12]: remove this line

File: app/javascript/controllers/shared/flash_controller.js [hunk 7]
```diff
@@ -5,7 +5,13 @@ export default class extends Controller {
...
+    this.alertTarget.style.display = 'none';
+    // Alternatively, if you want to remove the element from the DOM
```
Comment (jacky-le-goldenowl) [inline app/javascript/controllers/shared/flash_controller.js:14]: remove this 2 lines

File: app/javascript/pack/custom.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
+document.addEventListener('turbolinks:load', function(){
```
Comment (jacky-le-goldenowl) [inline app/javascript/pack/custom.js:1]: Move all code from pack to stimulus controller

File: app/models/room.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+#  user_id             :bigint           not null
+#
+# Indexes
+#
+#  index_rooms_on_renter_id  (renter_id)
+#  index_rooms_on_user_id    (user_id)
+#
+class Room < ApplicationRecord
+  belongs_to :user
+  belongs_to :renter, optional: true
```
Comment (jacky-le-goldenowl) [inline app/models/room.rb:31]: Should use has_one

File: app/models/room_service.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+class RoomsService < ApplicationRecord
```
Comment (jacky-le-goldenowl) [inline app/models/room_service.rb:3]: Set name model wrong

File: app/models/service.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+#
+#  id            :bigint           not null, primary key
+#  name          :string
+#  note          :text
+#  price_service :bigint
+#  created_at    :datetime         not null
+#  updated_at    :datetime         not null
+#
+class Service < ApplicationRecord
+  has_many :rooms_services, dependent: :destroy, class_name: 'RoomsService'
```
Comment (jacky-le-goldenowl) [inline app/models/service.rb:15]: remove this: ` class_name: 'RoomsService'`

File: app/models/user.rb [hunk 12]
```diff
@@ -47,6 +47,8 @@ class User < ApplicationRecord
          :omniauthable, omniauth_providers: [:google_oauth2]
   after_create :assign_default_role
 
+  has_many :rooms, dependent: :destroy
+
   def self.from_omniauth(auth)
     # create a new user with provider and uid followed by google to add to the database.
```
Comment (jacky-le-goldenowl) [inline app/models/user.rb:53]: Remove comment

File: app/views/layouts/application.html.erb [hunk 13]
```diff
@@ -12,6 +12,7 @@
     <%= vite_client_tag %>
     <%= vite_javascript_tag 'application' %>
     <%= favicon_link_tag 'logo/logo.png' %>
+    <%= javascript_include_tag '/pack/custom', "data-turbo-track": "reload", defer: true %>
```
Comment (jacky-le-goldenowl) [inline app/views/layouts/application.html.erb:15]: Remove this line, We just use stimulus to mange javascript code

File: app/views/renters/_form.html.erb [hunk 14]
```diff
@@ -0,0 +1,18 @@
+<%= simple_form_for @renter, method: (@renter.persisted? ? :put : :post), url: (@renter.persisted? ? users_renter_path : users_renters_path ) do |f| %>
+  <%= f.error_notification %>
+
+  <%= f.input :name, as: :string, wrapper_html: {class: 'mb-4'}, label: 'Name:', label_html: {class: 'mb-2 fw-bold'}, required: true, autofocus: true%>
+  <%= f.input :phone_number, as: :string, wrapper_html: {class: 'mb-4'}, label: "Phone Number:", label_html: {class: 'mb-2 fw-bold'}%>
+  <%= f.input :identity, as: :string, wrapper_html: {class: 'mb-4'}, label: "Identity:", label_html: {class: 'mb-2 fw-bold'} %>
+  <%= f.input :address, as: :string, wrapper_html: {class: 'mb-4'}, label: "Address: ", label_html: {class: 'mb-2 fw-bold'}%>
+  <%= f.input :gender, as: :radio_buttons, wrapper_html: {class: 'mb-4 flex'}, collection_html: {class: 'flex p-4'}, label_html: {class: 'mb-2 text-white'}, collection: [['0', 'female'], ['1', 'male']], label_method: :second, value_method: :second, checked: (@renter&.gender? ? 'male': 'female') %>
+
+ <%= f.simple_fields_for :rooms do |room_fields| %>
```
Comment (jacky-le-goldenowl) [inline app/views/renters/_form.html.erb:10]: I think we do not add room to renter when create

File: app/views/renters/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+<div class="main-container" style="background-color: wheat;" >
+        <div id="left-div" class="left-div" style="background-color: wheat;">
+            <div class="left-div-child" style="width: 100%;">
+                <%= render 'users/navigation'%>
+            </div>
+        </div>
+        <div id="right-container" class="right-container">
+            <%= render 'users/topnav'%>
+            <div class="right-div"
+                style="background-color: rgb(232, 247, 253); display: flex; flex-direction: column; justify-content: space-around; width: 100%; flex-wrap: wrap;">
```
Comment (jacky-le-goldenowl) [inline app/views/renters/index.html.erb:11]: Dont use style inline. Set class name and move all css style to scss file

File: app/views/renters/index.html.erb [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+                            <thead
+                                class="text-xs text-gray-700 uppercase bg-gray-50 dark:bg-gray-700 dark:text-gray-400">
+                                <tr>
+                                  <th scope="col" class="px-6 py-3">
+                                      Renter Name
+                                  </th>
+                                  <th scope="col" class="px-6 py-3">
+                                      <div class="flex items-center">
+                                          Phone Number
+                                          <a href="#"><svg class="w-3 h-3 ms-1.5" aria-hidden="true"
```
Comment (jacky-le-goldenowl) [inline app/views/renters/index.html.erb:40]: Move svg to svg image and call image with vite_image_tag

File: app/views/renters/index.html.erb [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+                                      <%= renter.phone_number%>
+                                    </td>
+                                    <td class="px-6 py-4">
+                                      <%= renter.identity%>
+                                    </td>
+                                    <td class="px-6 py-4">
+                                      <%= renter.address%>
+                                    </td>
+                                    <td class="px-6 py-4">
+                                      <%= renter.gender? ? "Nam" : "Nữ"%>
```
Comment (jacky-le-goldenowl) [inline app/views/renters/index.html.erb:132]: Should use gender type string

File: config/routes.rb [hunk 18]
```diff
@@ -17,6 +17,18 @@
     root 'dashboard#index'
   end
 
+  resource :users do
```
Comment (jacky-le-goldenowl) [inline config/routes.rb:20]: resource or resources ?

**Review by jacky-le-goldenowl — CHANGES_REQUESTED** [review]

**Comment by maironald** [comment]
I have changed:
- The room-controller, room-service
- Change css to scss
- Change the svg to image_tag
- Change javascript to the stimulus

**Pull request merged**